csharp_dotnetcore/17.Proactive-Messages/ProactiveBot.cs

```diff
@@ -0,0 +1,208 @@
+                                    $"| {j.TimeStamp} &nbsp; | {j.Conversation.Conversation.Id} &nbsp; | {j.Completed} |")));
+                        }
+                        else
+                        {
+                            await turnContext.SendActivityAsync("The job log is empty.");
+                        }
+
+                        break;
+
+                    default:
+
+                        // Check whether this is simulating a job completed event.
+                        string[] parts = text?.Split(' ', StringSplitOptions.RemoveEmptyEntries);
+                        if (parts != null && parts.Length is 2
+                            && parts[0].Equals("done", StringComparison.InvariantCultureIgnoreCase)
+                            && long.TryParse(parts[1], out long jobNumber))
+                        {
+                            if (!jobLog.TryGetValue(jobNumber, out JobLog.JobData jobInfo))
+                            {
+                                await turnContext.SendActivityAsync($"The log does not contain a job {jobInfo.TimeStamp}.");
+                            }
+                            else if (jobInfo.Completed)
+                            {
+                                await turnContext.SendActivityAsync($"Job {jobInfo.TimeStamp} is already complete.");
+                            }
+                            else
+                            {
+                                await turnContext.SendActivityAsync($"Completing job {jobInfo.TimeStamp}.");
+
+                                await CompleteJobAsync(turnContext.Adapter, AppId, jobInfo);
+                            }
+                        }
+
+                        break;
+                }
+
+                if (!turnContext.Responded)
+                {
+                    await turnContext.SendActivityAsync(
+                        "Type `run` or `run job` to start a new job.<br>" +
+                        "Type `show` or `show jobs` to display the job log.<br>" +
+                        "Type `done <jobNumber>` to complete a job.");
+                }
+            }
+        }
+
+        /// <summary>Handles non-message activities.</summary>
+        /// <param name="turnContext">The context object for this turn.</param>
+        /// <returns>A task that represents the work queued to execute.</returns>
+        private async Task OnSystemActivityAsync(ITurnContext turnContext)
+        {
+            // On a job completed event, mark the job as complete and notify the user.
+            if (turnContext.Activity.Type is ActivityTypes.Event)
+            {
+                var jobLog = await StateAccessors.JobLogData.GetAsync(turnContext, () => new JobLog());
+                var activity = turnContext.Activity.AsEventActivity();
+                if (activity.Name is JobCompleteEventName
+                    && activity.Value is long timestamp
```

Comment:
This is a very complex if statement. Consider breaking it down.

Reply:
Left it, alternate representations looked worse ;)